docs/architecture/adr-070-hasher.md

```diff
@@ -0,0 +1,146 @@
+
+```go
+package tmhash
+
+type Hasher interface {
+    hash.Hash
+    SumTruncated(bz []byte) []byte
+}
+```
+
+We would like to define the HashType for using in the later struct.
+```go
+type HashType string
+const (
+    SHA256 HashType = "sha256"
+    BLAKE2B HashType = "blake2b"
+    CUSTOM HashType = "custom"
+    // SHA256_SIMD HashType = "sha256_simd"
+    // other hash type if the TM core want to support
+)
+
+```
+
+The `HasherContainer` provide a container to keep the different `Hasher` instance base on the information
+provides when construcing the node.
+
+```go
+type HasherContainer struct {
+    hasherMap map[HashType]Hasher
+    RegisterType(HashType) error
+    GetHasher(HashType) (Hasher, error)
+}
+
+func (hc HasherContainer) RegisterType(ht HashType, customHasher Hasher) error {
+    switch ht {
+        case string(tmHash.SHA256):
+            hasherMap[ht] := sha256.New()
+            return nil
+        case string(tmHash.BLAKE2B):
+            hasherMap[ht] := blake2b.New()
+            return nil
+        case string(tmHash.CUSTOM):
+            hasherMap[ht] := customhash.New(customHasher)
+            return nil
+        default:
+            return errors.New("register an invalid hash type %w", ht) 
+    }
+}
+
+func (hc HasherContainer) GetHasher(ht HashType) Hasher {
+    return hc.hasherMap[ht]
+}
+
+```
+
+### Customizable hashing
+Because of we introduce a customize HashType - `CUSTOM`. The developer will need to implement their customize hasher and register it into the `HasherContainer` by calling `RegisterType(CUSTOM, customHasher)`
+
+```go
+package tmhash
+
+type CustomHash struct {
+    hasher tmHash.Hasher
+}
+
+var _ tmHash.Hasher = (*CustomHash)(nil)
+
+func New(hr tmhash.Hasher) tmhash.Hasher {
+    return &CustomHash{hasher: hr}
+}
+
+func (ch *CustomHash) SumTruncated(bz []byte) []byte {
+	return ch.hasher.SumTruncated(bz)
+}
+
+// other implementation for the hash.Hash functions
+...
+``` 
+
+### Chain parameter settings
+The hash algorithm is part of the consensus parameters. Therefore, we need to add
+new object `hash` into `genesis.json`. There are 2 objects in the `hash` object. `types` will be the hash algorithm be used in the hasher; The `custom_tx_hash_enable` allows the transaction to return its hash using custom hasher. The default value of `types` will be `sha256` and `custom_tx_hash_enable` will be `false` For compatibility, the current `genesis.json` without object `hash` settings will be using the default value above.
+
+```json
+{   "chain_id": "tm-chain",
+    ...
+    "consensus_params": {
+        "block": {...},
+        "evidence": {...},
+        "validator": {...},
+        "version": {...},
+        "hash": {
+            "types": "sha256",
+            "custom_tx_hash_enable": "true"
+        }
```

Comment:
does it have to be a genesis param? It should be enough if it's an app param - the node will not sync if it's using different setting than the rest of the network.

Reply:
however, external tools will need to know the hashing algorithm as well (eg to monitor TX confirmation), so having it in the genesis makes it more accessible. WDYT?

Reply:
> does it have to be a genesis param? It should be enough if it's an app param - the node will not sync if it's using different setting than the rest of the network.

Is it okay to put it into the app param? In my mind, it's a consensus-bound param.

> however, external tools will need to know the hashing algorithm as well (eg to monitor TX confirmation), so having it in the genesis makes it more accessible. WDYT?

Didn't think about the tooling level. but I agree, if the monitor/explorers are able to utilize it, it might make the tooling easier to connect to the TM core projects?

Reply:
yes

Reply:
I don't think this should be a consensus parameter. This would imply that it could change on-chain which would imply we need to build logic to handle the case of the hasher changing. We can add it to the genesis file though and expose what hasher is used via the RPC and app's gRPC.

Reply:
would be good to introduce genesis params #6814

Reply:
BTW: by app param I meant an app option - not a x/param thing.